### PR TITLE
Optimize and balance the farmer.

### LINKED ISSUE #236: Farmer more balanced
Make the farmer more balanced.

IE. Higher levels = more fields, faster actions.

### PR DESCRIPTION
**Pull request by Dadadah**
Closes #236
# Changes proposed in this pull request:
- Make farmer till the field when it is new.
- Give farmer a delay between each action of WORK

Review please @Minecolonies/maintainers

### CODE REVIEW

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 1]
```diff
@@ -41,8 +41,8 @@
      */
     private static final int     DELAY_DIVIDER       = 10;
     /**
-     * The EXP Earned per harvest.
-     */
+      * The EXP Earned per harvest.
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: why are you changing this back? Do you use our formatter?

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ private AIState prepareForFarming()
 
         if (currentField.needsWork())
         {
-            if (!checkForHoe() && canGoPlanting(currentField, building))
+            if(!checkForHoe() && canGoPlanting(currentField, building))
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: Definitely import our formater in the documentation folder!

Comment (Oggal) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: That missing space is on me, habit of removing it when IDE's add it that I picked up in languages that would flip for having it.

Comment (Kostronor) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: I suggest using our formatter and formatting with that ;)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 3]
```diff
@@ -537,4 +514,46 @@ public EntityCitizen getCitizen()
...
         return worker;
     }
+
+    /**
+      * This method allows us to harvest crops and leave the plant there.
+      * Credit goes to RightClickHarvest mod.
+      *
+      * @params position the position of the crop to harvest
+      */
+    private boolean harvestCrop(final BlockPos position)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Remove this unused private "harvestCrop" method. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 4]
```diff
@@ -258,24 +277,82 @@ private AIState cycle()
     }
 
     /**
-     * Checks if we can harvest, and does so if we can.
-     */
-    private void harvestIfAble(final BlockPos position)
+      * This (re)initializes a field.
+      * Checks the block above to see if it is a plant, if so, breaks it. Then tills.
+      */
+    private AIState initialize()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "initialize" is 14 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L109](https://github.com/Minecolonies/minecolonies/blob/d66a7520cb55dfd6009d5e5d65da0c0db70db8be/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L109): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L109](https://github.com/Minecolonies/minecolonies/blob/7b1bac02c876b79fa4b2623d9db171d01d5d2d2f/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L109): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/859f0e7392d9d3cf0b59502e4e663b46388424f7/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 5]
```diff
@@ -284,206 +420,216 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
     /**
      * Handles the offset of the field for the farmer.
+     * Skips to the next harvestable crop, returns true if one was found.
      *
      * @param field the field object.
-     * @return true if successful.
+     * @return true if a harvestable crop was found.
      */
-    private boolean handleOffset(@NotNull final Field field)
+    private boolean handleOffsetHarvest(@NotNull final Field field)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR The Cyclomatic Complexity of this method "handleOffsetHarvest" is 11 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/72f1d05b7b7c740273c982731d794c9369b519c7/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/143d6bdb2564326cecfc887ec78cb73f820faa1c/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/85cb92729b6f94689bbeaafbcd8a9d69ea9cc718/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 6]
```diff
@@ -284,206 +425,216 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR The Cyclomatic Complexity of this method "handleOffsetHarvest" is 11 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/ee6168034d857e5ad752af9a9114498725ced4aa/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 7]
```diff
@@ -284,206 +425,247 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
-    private void resetVariables()
+    private void equipHoe()
     {
-        requestSeeds = true;
-        shouldDumpInventory = true;
-        shouldTryToGetSeed = true;
+        worker.setHeldItem(getHoeSlot());
     }
+    
+    private int dist = 0;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1213)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 8]
```diff
@@ -284,206 +425,247 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
+    private boolean horizontal = false;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1213)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 9]
```diff
@@ -284,206 +425,247 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
+    private int totalDis = 1;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1213)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 10]
```diff
@@ -284,206 +425,247 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
         {
-            chatSpamFilter.talkWithoutSpam("entity.farmer.NeedSeed", field.getSeed().getItemStackDisplayName(new ItemStack(field.getSeed())));
+            workingOffset = new BlockPos(1, 0,0);
+            totalDis = 1;
+            dist = 0;
         }
         else
         {
-            buildingFarmer.getCurrentField().setNeedsWork(false);
+            final int absZ = Math.abs(workingOffset.getZ());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Remove this unused "absZ" local variable. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1481)
MAJOR Remove this useless assignment to local variable "absZ". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1854)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 11]
```diff
@@ -284,206 +425,247 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
+            if (Math.abs(workingOffset.getZ()) >= field.getWidthPlusZ() && Math.abs(workingOffset.getX()) >= field.getLengthPlusX())
+            {
+                System.out.println(field.getWidthPlusZ());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS106)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 12]
```diff
@@ -284,206 +425,247 @@ private void hoeIfAble(final BlockPos position, final Field field)
...
+                if (totalDis == dist)
+                {
+                 horizontal = !horizontal;
+                 dist = 0;
+                 if (horizontal)
+                 {
+                    totalDis++; 
+                 }
+                }
+                if (horizontal){
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 11 issues
* MAJOR 7 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/09e9b0d75c3bcfdbccfcdf442814acabaa9150b7/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L122](https://github.com/Minecolonies/minecolonies/blob/fa1d1de6d1f1cf36082cc7bdabc330853a940397/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L122): Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 13]
```diff
@@ -66,6 +78,13 @@
     private boolean shouldTryToGetSeed = true;
 
     /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L129](https://github.com/Minecolonies/minecolonies/blob/5780e56b3969383589cd183dc59ff69c28a62f9a/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L129): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 6 issues
* MAJOR 4 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L135](https://github.com/Minecolonies/minecolonies/blob/c707eb50ef822357443e74b12bb1fa8475dbf6cf/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L135): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 14]
```diff
@@ -22,10 +22,10 @@
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-import java.util.List;
-
 import static com.minecolonies.entity.ai.util.AIState.*;
 
+import java.util.List;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Remove this unused import 'java.util.List'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 9 issues
* MAJOR 4 major
* MINOR 5 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L121](https://github.com/Minecolonies/minecolonies/blob/4f77c95eed0a3f1e77cc7e11a6e50ca0092f6694/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L121): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L123](https://github.com/Minecolonies/minecolonies/blob/cbd272bbe193daaf1b44412f8e290a7cef13970b/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L123): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 15]
```diff
@@ -21,550 +21,698 @@
...
+            return AIState.IDLE;
+        }
+        return AIState.FARMER_WORK;
+    }
+
+    /**
+     * This (re)initializes a field.
+     * Checks the block above to see if it is a plant, if so, breaks it. Then tills.
+     */
+    private AIState initialize()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**